sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -66,980 +85,305 @@ case class ScalaUDF(
-    val catalystConverterTerm = ctx.freshName("catalystConverter")
-    val catalystConverterTermIdx = ctx.references.size - 1
-    ctx.addMutableState(converterClassName, catalystConverterTerm,
-      s"this.$catalystConverterTerm = ($converterClassName)$typeConvertersClassName" +
-        s".createToCatalystConverter((($scalaUDFClassName)expressions" +
-          s"[$catalystConverterTermIdx]).dataType());")
+    val expressionEncoderClassName = classOf[ExpressionEncoder[Row]].getName
+    val rowEncoderClassName = RowEncoder.getClass.getName + ".MODULE$"
+    val structTypeClassName = StructType.getClass.getName + ".MODULE$"
+    val rowClassName = Row.getClass.getName + ".MODULE$"
+    val rowClass = classOf[Row].getName
+    val internalRowClassName = classOf[InternalRow].getName
+    // scalastyle:off
+    val javaConversionClassName = scala.collection.JavaConversions.getClass.getName + ".MODULE$"
+    // scalastyle:on
+
+    // Generate code for input encoder
+    val inputExpressionEncoderTerm = ctx.freshName("inputExpressionEncoder")
+    ctx.addMutableState(expressionEncoderClassName, inputExpressionEncoderTerm,
+      s"this.$inputExpressionEncoderTerm = ($expressionEncoderClassName)$rowEncoderClassName" +
+        s".apply((($scalaUDFClassName)expressions" +
+          s"[$scalaUDFTermIdx]).getInputSchema());")
+
+    // Generate code for output encoder
+    val outputExpressionEncoderTerm = ctx.freshName("outputExpressionEncoder")
+    ctx.addMutableState(expressionEncoderClassName, outputExpressionEncoderTerm,
+      s"this.$outputExpressionEncoderTerm = ($expressionEncoderClassName)$rowEncoderClassName" +
+        s".apply((($scalaUDFClassName)expressions[$scalaUDFTermIdx]).getDataType());")
 
     val resultTerm = ctx.freshName("result")
 
-    // This must be called before children expressions' codegen
-    // because ctx.references is used in genCodeForConverter
-    val converterTerms = (0 until children.size).map(genCodeForConverter(ctx, _))
-
     // Initialize user-defined function
     val funcClassName = s"scala.Function${children.size}"
 
     val funcTerm = ctx.freshName("udf")
-    val funcExpressionIdx = ctx.references.size - 1
     ctx.addMutableState(funcClassName, funcTerm,
       s"this.$funcTerm = ($funcClassName)((($scalaUDFClassName)expressions" +
-        s"[$funcExpressionIdx]).userDefinedFunc());")
+        s"[$scalaUDFTermIdx]).userDefinedFunc());")
 
     // codegen for children expressions
     val evals = children.map(_.gen(ctx))
+    val evalsArgs = evals.map(_.value).mkString(", ")
+    val evalsAsSeq = s"$javaConversionClassName.asScalaIterable" +
+      s"(java.util.Arrays.asList($evalsArgs)).toList()"
+    val inputInternalRowTerm = ctx.freshName("inputRow")
+    val inputInternalRow = s"$rowClass $inputInternalRowTerm = " +
+      s"($rowClass)$inputExpressionEncoderTerm.fromRow(InternalRow.fromSeq($evalsAsSeq));"
```

Comment:
The reason to construct a row here is to encode expression results to scala objects using row encoder. Because UDF accepts and processes scala objects, we need to encode its input and decode its output.